table/tables/cache_test.go

```diff
@@ -201,3 +202,36 @@ func TestCacheTableBasicReadAndWrite(t *testing.T) {
 	}
 	tk.MustQuery("select *from write_tmp1").Check(testkit.Rows("1 101 1001", "2 222 3333", "3 113 1003"))
 }
+
+func TestCacheTableComplexRead(t *testing.T) {
+	t.Parallel()
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+	doneCh := make(chan struct{}, 1)
+	tk1 := testkit.NewTestKit(t, store)
+	tk2 := testkit.NewTestKit(t, store)
+	tk1.MustExec("use test")
+	tk2.MustExec("use test")
+	tk1.MustExec("create table complex_cache (id int primary key auto_increment, u int unique, v int)")
+	tk1.MustExec("insert into complex_cache values" + "(5, 105, 1005), (7, 117, 1007), (9, 109, 1009)")
+	tk1.MustExec("alter table complex_cache cache")
+	tk1.MustExec("begin")
+	tk1.MustQuery("select *from complex_cache where id > 7").Check(testkit.Rows("9 109 1009"))
+
+	go func() {
+		tk2.MustExec("begin")
+		tk2.MustQuery("select *from complex_cache where id > 7").Check(testkit.Rows("9 109 1009"))
+		var i int
+		for i = 0; i < 10; i++ {
+			time.Sleep(100 * time.Millisecond)
+			if tk2.HasPlan("select *from complex_cache where id > 7", "UnionScan") {
+				break
+			}
```

Comment:
Fail if still can't get the plan after several retries.

Reply:
okay